File: section2/103.二叉树的锯齿形层序遍历.py
#
# @lc app=leetcode.cn id=103 lang=python3
#
# [103] 二叉树的锯齿形层序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __str__(self):
        return f'node(val={self.val})'

    @staticmethod
    def generate(nums: list):
        root = TreeNode(nums[0])
        nodes = [root]

        index = 1
        while index < len(nums):
            new_nodes = []
            for node in nodes:
                if node:
                    node.left = None if nums[index] is None else TreeNode(nums[index])
                    node.right = None if nums[index + 1] is None else TreeNode(nums[index + 1])

                new_nodes.append(node.left if node else None)
                new_nodes.append(node.right if node else None)
                index += 2

            nodes = new_nodes

        return root


class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        stacks = [root]
        rows = []
        layer = 1
        while stacks:
            row, new_stacks = [], []
            for node in stacks:
                if node:
                    row.append(node.val)
                    new_stacks.append(node.left)
                    new_stacks.append(node.right)

            stacks = new_stacks
            if layer % 2 == 0:
                row = list(reversed(row))

            if row:
                rows.append(row)

            layer += 1

        return rows

    def test(self):
        # [
        #     [3],
        #     [20,9],
        #     [15,7]
        # ]
        print(self.zigzagLevelOrder(TreeNode.generate([3, 9, 20, None, None, 15, 7])))


# Solution().test()
# @lc code=end

File: section2/section/103.二叉树的锯齿形层序遍历.go
package section

import "fmt"

/*
 * @lc app=leetcode.cn id=103 lang=golang
 *
 * [103] 二叉树的锯齿形层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return nil
	}

	// 收集所有的数据
	values := make([][]int, 0)
	// 当前行所有的节点，初始化的时候为第一行 root 节点数据
	currents := []*TreeNode{root}
	// 方向控制，默认正向遍历
	position := true
	for len(currents) > 0 {
		currentValues := make([]int, 0)

		var nextCurrents []*TreeNode
		// 遍历所有 root 节点
		for index, currentNode := range currents {
			// 如果是负向遍历的时候，将 index 改成从尾部开始
			if !position {
				index = len(currents) - index - 1
			}

			// 收集节点数据
			currentValues = append(currentValues, currents[index].Val)
			// 将子节点放到下一次遍历数据中
			if currentNode.Left != nil {
				nextCurrents = append(nextCurrents, currentNode.Left)
			}

			if currentNode.Right != nil {
				nextCurrents = append(nextCurrents, currentNode.Right)
			}
		}

		values = append(values, currentValues)
		currents = nextCurrents
		// 重置方向
		position = !position
	}

	return values
}

func Test103() {
	// [[3],[20,9],[15,7]]
	fmt.Println(zigzagLevelOrder(ToNode([]int{3, 9, 20, 0, 0, 15, 7})))

	// [[1]]
	fmt.Println(zigzagLevelOrder(ToNode([]int{1})))

	// []
	fmt.Println(zigzagLevelOrder(ToNode([]int{})))
}

// @lc code=end
